src/features.c

```diff
@@ -0,0 +1,171 @@
+/*
+ * crun - OCI runtime written in C
+ *
+ * crun is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * crun is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with crun.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <config.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <argp.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <yajl/yajl_tree.h>
+#include <yajl/yajl_gen.h>
+
+#include "crun.h"
+#include "libcrun/container.h"
+#include "libcrun/utils.h"
+
+static char doc[] = "OCI runtime";
+
+static struct argp_option options[] = { {0} };
+
+static char args_doc[] = "features";
+
+static error_t
+parse_opt (int key, char *arg arg_unused, struct argp_state *state arg_unused)
+{
+  if (key != ARGP_KEY_NO_ARGS) {
+        return ARGP_ERR_UNKNOWN;
+    }
+
+    return 0;
+}
+
+static struct argp run_argp = { options, parse_opt, args_doc, doc, NULL, NULL, NULL };
+
+void 
+addStringToJSON(yajl_gen jsonGen, const char *key, const char *value) {
+    yajl_gen_string(jsonGen, (const unsigned char *) key, strlen(key));
+    yajl_gen_string(jsonGen, (const unsigned char *) value, strlen(value));
+}
+
+void 
+addBoolToJSON(yajl_gen jsonGen, const char *key, int value) {
+    yajl_gen_string(jsonGen, (const unsigned char *) key, strlen(key));
+    yajl_gen_bool(jsonGen, value);
+}
+
+void 
+addArrayToJSON(yajl_gen jsonGen, const char *key, const char **array) {
+    yajl_gen_string(jsonGen, (const unsigned char *) key, strlen(key));
+    yajl_gen_array_open(jsonGen);
+
+    for (size_t i = 0; array[i] != NULL; i++) {
+        yajl_gen_string(jsonGen, (const unsigned char *) array[i], strlen(array[i]));
+    }
+
+    yajl_gen_array_close(jsonGen);
+}
+
+int 
+crun_command_features(struct crun_global_arguments *global_args, int argc, char **argv, libcrun_error_t *err) {
+    argp_parse(&run_argp, argc, argv, 0, 0, &options);
+
+    // Call the function in features.c to gather the feature information
+    struct features_info_s info;
```

Comment:
let's change the code to allocate the struct as:

```
struct features_info_s *info;
int result = libcrun_container_print_features(&info);
```

and have a function that cleans it up, since there are many allocations that are just leaked at the moment.

Then we can define this feature with a cleanup annotation:

```
cleanup_struct_features struct features_info_s *info = NULL;
```

that will automatically call the cleanup function when the pointer is not NULL.  You can look at other `cleanup_` functions defined in `libcrun/utils.h`